### DISCUSSION TITLE
How to add ts type/interface to custom request?

### DISCUSSION THREAD

**Discussion by mblachlov24** [nuxt/nuxt#16702] [Questions]
It works fine here:

```ts
interface Response {
 name: string,
}

const { data } = await useAsyncData<Response[]>(() => $fetch('/api/whatever'))
```

However, when I try this on my own $fetch instance it doesn't work.

```ts
interface Response {
    name: string,
}

const { data } = await useAsyncData<Response[]>(() => useRequest('/api/whatever'))
```

_const useApi: (path: string, options?: any) => Promise<unknown>
Type 'Promise<unknown>' is not assignable to type 'Promise<Response[]>'.
  Type '{}' is missing the following properties from type 'Response[]': length, pop, push, concat, and 29 more.ts(2322)
asyncData.d.ts(27, 193): The expected type comes from the return type of this signature._

useRequest.ts
```ts
export const useRequest = (path: string, options?: any) => {
    const config = useRuntimeConfig()
    const token = useCookie('XSRF-TOKEN')

    return $fetch(path, {
        baseURL: config.public.apiUrl,
        credentials: 'include',
        headers: {
            Accept: 'application/json',
            'X-XSRF-TOKEN': token.value,
        },
        ...options
    })
}
```

**Comment by bloomlive** [comment]
This seems to be problem in my case as well and because have quite a lot of headers in our case.

Temporarily I'm doing the following

```typescript
const { centralApi } = useApi()
```

```typescript
const messages = ref(
  (data.value as INotificationResponse[]).map((item) => {
    return {
      ...item.data,
      time: item.created_at,
    }
  })
)
```

**Comment by RSS1102** [comment]
I have the same problem. https://github.com/nuxt/framework/discussions/7936

**Reply by RSS1102** [reply]
> 我也有同样的问题。#7936

![image](https://user-images.githubusercontent.com/81673017/193356752-eef82d94-b2af-40c5-9857-c41169b0c93e.png)
I got inspiration from you. It really works here
